rfcs/007-bytecode-interpreter.md

```diff
@@ -0,0 +1,839 @@
+reduced to rather low-level operation: environment manipulation, function
+application, boolean and integers operations, branching, exceptions, memory
+allocation/value creation.
+
+While the ZAM is considered stack-based, it has a few predefined registers:
+
+- PC: program counter
+- SP: stack pointer
+- ACCU: accumulator
+- TRAPSP: stack pointer of highest exception handler
+- EXTRA_ARGS: number of extra arguments to function application
+- ENV: environment
+- GLOBAL: global data
+
+#### References
+
+- [Real World OCaml chapter on bytecode](https://dev.realworldocaml.org/compiler-backend.html),
+- [Original Leroy's paper: _The ZINC experiment: an economical implementation of
+  the ML language_ (1990)](https://inria.hal.science/inria-00070049/document)
+- [OCaml bytecode instructions](http://cadmium.x9c.fr/distrib/caml-instructions.pdf)
+- [OCaml memory representation of values](https://ocaml.org/docs/memory-representation)
+- [From Krivine's machine to the Caml implementations](https://xavierleroy.org/talks/zam-kazam05.pdf)
+- [OCaml's representation of closures](https://github.com/Gbury/ocaml-memgraph/blob/master/doc/closures.md)
+
+### Haskell
+
+Despite not being advertised, Haskell has an interpreter as well, which is used
+mostly for the GHCi REPL. This section describes what we know of the actual the
+bytecode interpreter, but also incorporates some aspects of the original STG
+machine which is the basis of the low-level operational semantics of Haskell.
+
+#### Memory representation
+
+Closures (thunks and functions) are represented uniformly as an environment and
+a pointer to a closure info table. The closure info table contains the code
+pointer and other garbage-collection related data. The closure info table is
+generated only once per binding appearing in the code and thus shared by dynamic
+instances of the same code. In some other machines for lazy lambda calculi,
+thunks hold additional data such as a flag (evaluated or suspended) and a
+potential value that is filled after the thunk has been forced. When accessing a
+thunk, the caller checks the state, and then either retrieve the value or
+initiate an update sequence and retrieve the code of the unevaluated expression.
+In Haskell, the update process is performed _by the callee_ instead (the thunk's
+code), such that thunk access is uniform: it's an unconditional jump to the
+corresponding code.
+
+As with other VMs, the environment -- immediately inlined after the code pointer
+-- stores the free variables of the expression that are captured by the closure.
+When entering a closure, a dedicated register holds the pointer to the
+environment for fast access.
+
+The STG paper argues that this uniform thunk representation (with "self-handled
+update") simplifies the compilation process and gives room for some specific
+optimizations (vectored return for pattern matching, for example) that should be
+beneficial to Haskell programs.
```

Comment:
It's worth noting that since the STG paper was written, things have become a little more complex. GHC uses some pointer tagging to mark if the thunk is already forced in the form of one of the 3 (on 32 bits) or 7 (on 64 bits) constructors of the data type. So pattern-matching will check these bits before entering the closure. For efficiency.